uv/manager/manager.py

```diff
@@ -0,0 +1,101 @@
+  particular reporter, which it will use to report whatever measurements is
+  conducts.
+
+  The MeasurementManager has a set of measurements---stored in a dictionary
+  (described below)---which are to be carried out at a certain frequency, which
+  is specific to each measurement.
+
+  Whenever the MeasurementManager's process() function is called, with a
+  step number provided as an argument, the MeasurementManager decides which
+  measurements are required at each step, carries them out, and reports them.
+
+  In order to carry out the measurement, the Manager has a state (Dict) at each
+  step, which contains all the information necessary to perform the measurement.
+  Part of this state is static (e.g., the training dataset), while part can
+  change at each step.  This dynamic state is passed along in the process() call.
+
+  Args:
+    static_state: Dict of variables which will be used to compute the measured
+                  values.  The static_state is fixed throughout time.
+    reporter: Reporter which will report the measurements.
+  """
+
+  def __init__(self, static_state: Dict[str, Any],
+               reporter: Type[AbstractReporter]):
+    self.static_state = static_state
+    self.reporter = reporter
+    self.measurements = {}
+
+  def add_measurement(self, measurement_spec: Dict[str, Union[str, int,
+                                                              Callable]]):
+    """Adds a measurement specification to the MeasurementManager.
+    Args:
+      measurement_spec: Dict containing keys:
+        'name': String, the name of the measurement, e.g. 'training_loss'
+        'interval': Integer, how often to measure
+        'function': Function, taking in the state_dictionary, and returning the
+                              measured value
+    """
+    name = measurement_spec['name']
+    self.measurements[name] = {
+        'interval': measurement_spec['interval'],
+        'function': measurement_spec['function']
+    }
+
+  def process(self, step: int, dynamic_state: Dict[str, Any]):
+    """ At a given step, decide which measurements need to be
+    performed at this step and perform them
+
+    Args:
+      step: Integer, the number of the current step
+      dynamic_stat: Dict, the state variables necessary to perform the
+                    measurement
+    """
+
+    to_measure = []
+    for msmt_name, msmt_spec in self.measurements.items():
+      if step % msmt_spec['interval'] == 0:
+        to_measure.append(msmt_name)
+
+    if len(to_measure) > 0:
+      self.trigger_subset(step, dynamic_state, to_measure)
+
+  def trigger_subset(self, step: int, dynamic_state: Dict[str, Any],
+                     msmt_list: Iterable[str]):
+    """ Regardless of step, performs the measurements specified
+    by msmt_list and reports them """
+
+    full_state = {**self.static_state, **dynamic_state}
+
+    msmts = {}
+    for msmt_name in msmt_list:
+      msmt_fun = self.measurements[msmt_name]['function']
```

Comment:
`fn` is usually the function abbreviation vs `fun`

Reply:
ty 👍